presto-main/src/main/java/com/facebook/presto/operator/SpatialJoinOperator.java

```diff
@@ -0,0 +1,272 @@
+                List<Type> probeTypes,
+                List<Integer> probeOutputChannels,
+                int probeGeometryChannel,
+                PagesSpatialIndexFactory pagesSpatialIndexFactory)
+        {
+            this.operatorId = operatorId;
+            this.planNodeId = planNodeId;
+            this.probeTypes = ImmutableList.copyOf(probeTypes);
+            this.probeOutputTypes = probeOutputChannels.stream()
+                    .map(probeTypes::get)
+                    .collect(toImmutableList());
+            this.buildOutputTypes = pagesSpatialIndexFactory.getOutputTypes();
+            this.probeOutputChannels = ImmutableList.copyOf(probeOutputChannels);
+            this.probeGeometryChannel = probeGeometryChannel;
+            this.pagesSpatialIndexFactory = pagesSpatialIndexFactory;
+        }
+
+        @Override
+        public List<Type> getTypes()
+        {
+            return ImmutableList.<Type>builder()
+                    .addAll(probeOutputTypes)
+                    .addAll(buildOutputTypes)
+                    .build();
+        }
+
+        @Override
+        public Operator createOperator(DriverContext driverContext)
+        {
+            checkState(!closed, "Factory is already closed");
+            OperatorContext operatorContext = driverContext.addOperatorContext(
+                    operatorId,
+                    planNodeId,
+                    SpatialJoinOperator.class.getSimpleName());
+            return new SpatialJoinOperator(
+                    operatorContext,
+                    getTypes(),
+                    probeTypes,
+                    probeOutputChannels,
+                    probeGeometryChannel,
+                    pagesSpatialIndexFactory);
+        }
+
+        @Override
+        public void noMoreOperators()
+        {
+            if (closed) {
+                return;
+            }
+
+            pagesSpatialIndexFactory.noMoreProbeOperators();
+            closed = true;
+        }
+
+        @Override
+        public OperatorFactory duplicate()
+        {
+            return new SpatialJoinOperatorFactory(operatorId, planNodeId, probeTypes, probeOutputChannels, probeGeometryChannel, pagesSpatialIndexFactory);
+        }
+    }
+
+    private final OperatorContext operatorContext;
+    private final List<Type> probeTypes;
+    private final List<Type> outputTypes;
+    private final List<Integer> probeOutputChannels;
+    private final int probeGeometryChannel;
+    private final PagesSpatialIndexFactory pagesSpatialIndexFactory;
+
+    private ListenableFuture<PagesSpatialIndex> pagesSpatialIndexFuture;
+    private final PageBuilder pageBuilder;
+    @Nullable
+    private Page probe;
+
+    // The following fields represent the state of the operator in case when processProbe yielded or
+    // filled up pageBuilder before processing all records in a probe page.
+    private int probePosition;
+    @Nullable
+    private long[] joinAddresses;
```

Comment:
Can this be significant in size (e.g. more than size of a page)? Should this be accounted for?

Reply:
@findepi In practice, I don't think this can be large. This array holds addresses of records on the build side that match a single record on the probe side.

Reply:
I am aware. Pessimistically, this can be an array listing all positions on the right side. If that's not possible and there is some hard limit on the array, you can e.g `verify(joinAddresses.length < HARD_LIMIT)` in the code. If there isn't such a hard limit, you can easily account for this memory.

Reply:
@findepi Let me update the operator to account for this memory.